File: tier1_foundation/fast_slow_pointers/problems/p202_happy_number.c
/*
 * P202: Happy Number (Easy)
 * https://leetcode.com/problems/happy-number/
 * Topics: Hash Table, Math, Two Pointers
 * 
 * Write an algorithm to determine if a number n is happy.
 * A happy number is a number defined by the following process:
 * Return true if n is a happy number, and false if not.
 * 
 * Example 1:
 *     Input: n = 19
 *     Output: true
 *     Explanation:
 *     12 + 92 = 82
 *     82 + 22 = 68
 *     62 + 82 = 100
 *     12 + 02 + 02 = 1
 * 
 * Example 2:
 *     Input: n = 2
 *     Output: false
 * 
 * Constraints:
 *     - 1 <= n <= 231 - 1
 * 
 * Template (python3):
 *     class Solution:
 *         def isHappy(self, n: int) -> bool:
 * 
 * Hint: Use fast and slow pointers on the sequence of sum-of-squared-digits.
 */
#include "ctest.h"

static int get_next(int num)
{
    abort();
}

int isHappy(int n)
{
    abort();
}

void __attribute__((unused)) _use_harness_fns(void)
{
    abort();
}

int main(void)
{
    TestCase tests[] = {
        {.label = "example 1", .target = 19, .expected = {1}, .expected_n = 1},
        {.label = "example 2", .target = 2, .expected = {0}, .expected_n = 1},
        {.label = "already happy", .target = 1, .expected = {1}, .expected_n = 1},
        {.label = "small happy number", .target = 7, .expected = {1}, .expected_n = 1},
        {.label = "enters 4-16 cycle", .target = 4, .expected = {0}, .expected_n = 1},
        {.label = "power of 10", .target = 100, .expected = {1}, .expected_n = 1},
        {.label = "large unhappy number", .target = 999999999, .expected = {0}, .expected_n = 1},
    };
    int n = sizeof(tests) / sizeof(tests[0]);
    int passed = 0;
    for (int i = 0; i < n; i++) {
        int got = isHappy(tests[i].target);
        if (got == tests[i].expected[0]) {
            passed++;
            printf("  Test %d (%s): PASS\n", i + 1, tests[i].label);
        } else {
            printf("  Test %d (%s): FAIL\n", i + 1, tests[i].label);
            printf("    Expected: %d\n    Got:      %d\n",
                   tests[i].expected[0], got);
        }
    }
    printf("\n  %d/%d passed\n", passed, n);
    return passed == n ? 0 : 1;
}

File: tier1_foundation/fast_slow_pointers/solutions/p202_happy_number.mjs
/**
 * P202: Happy Number (Easy)
 * https://leetcode.com/problems/happy-number/
 * Topics: Hash Table, Math, Two Pointers
 *
 * Write an algorithm to determine if a number n is happy.
 * A happy number is a number defined by the following process:
 *
 * 1. Starting with any positive integer, replace the number by the sum of the squares of its digits.
 * 2. Repeat the process until the number equals 1 (where it will stay), or it loops endlessly in a cycle which does not include 1.
 * 3. Those numbers for which this process ends in 1 are happy.
 * Return true if n is a happy number, and false if not.
 *
 * Example 1:
 *     Input: n = 19
 *     Output: true
 *     Explanation:
 *     1^2 + 9^2 = 82
 *     8^2 + 2^2 = 68
 *     6^2 + 8^2 = 100
 *     1^2 + 0^2 + 0^2 = 1
 *
 * Example 2:
 *     Input: n = 2
 *     Output: false
 *
 * Constraints:
 *     - 1 <= n <= 2^31 - 1
 *
 * Template (python3):
 *     class Solution:
 *         def isHappy(self, n: int) -> bool:
 *
 * Hint: Use fast and slow pointers on the sequence of sum-of-squared-digits.
 */

import { readLine, readInts, readInt, writeInt, writeInts, writeString, writeBool } from '../../wasm_libs/js/io.mjs';

function getNext(num) {
  let total = 0;
  while (num > 0) {
    const digit = num % 10;
    total += digit * digit;
    num = Math.floor(num / 10);
  }
  return total;
}

function solve(n) {
  let slow = n;
  let fast = getNext(n);
  while (fast !== 1 && slow !== fast) {
    slow = getNext(slow);
    fast = getNext(getNext(fast));
  }
  return fast === 1;
}

const n = readInt();
writeBool(solve(n));
